upgrade_v3.php

```diff
@@ -0,0 +1,1974 @@
+                }
+                $html .= '</ul>';
+                $logData .= "\n";
+                writeToLog($logData);
+            }
+
+            if (count($preUpgradeCheckErrors) === 0 && count($preUpgradeCheckWarnings) > 0) {
+                // The checkbox doesn't do anything, but is just there to make users aware that they are doing risky things.
+                $html .= $generalRemarks . '
+                <div style="text-align: left">
+                    <input type="checkbox" id="forceUpgradeStart" /> <label for="forceUpgradeStart">Yes, I am aware of the warnings above and still want to proceed with the upgrade.</label>
+                </div><br />
+                <button class="btn btn-primary" id="startUpgradeButton">Start the upgrade</button>';
+            }
+
+            if (count($preUpgradeCheckErrors) === 0 && count($preUpgradeCheckWarnings) === 0) {
+                writeToLog("All pre-upgrade checks passed successfully.");
+                $html .= "<h3>Ready to upgrade ✅</h3>
+                <br />Your system is compatible with Mautic 3!<br>Do not refresh or stop the process. This may take several minutes.<br><br>
+                " . $generalRemarks . "
+                <button class=\"btn btn-primary\" id=\"startUpgradeButton\">Start the upgrade</button>";
+            }
+
+            outputJSON('success', $html);
+
+        case 'startUpgrade':
+            writeToLog("Starting the upgrade...");
+            sendUpgradeStats('started');
+            outputJSON('success', 'OK', 'backupDatabase', 'Backing up the database if we can...');
+
+        case 'backupDatabase':
+            // Only do the backup if mysqldump is available, otherwise skip this step.
+            if (is_mysqldump_available() === true) {
+                writeToLog("Running database backup...");
+                list($success, $message) = backup_database();
+
+                if (!$success) {
+                    throwErrorAndWriteToLog(
+                        "ERR_DATABASE_BACKUP_FAILED",
+                        "Database backup failed. Your Mautic 2 installation is intact, so you can safely restart the upgrade. Error from mysqldump: " . $message
+                    );
+                }
+                writeToLog("Successfully backed up database.");
+            } else {
+                writeToLog("Skipping database backup due to mysqldump not being available.");
+            }
+
+            outputJSON('success', 'OK', 'applyV2Migrations', 'Applying Mautic 2 database migrations to ensure all migrations are in place prior to upgrade...');
+
+        case 'applyV2Migrations':
+            writeToLog("Getting the amount of available Mautic 2 database migrations...");
+            // Get the amount of available database migrations
+            list($success, $data) = get_available_migrations();
+
+            if ($success === false) {
+                throwErrorAndWriteToLog(
+                    "ERR_MAUTIC_2_MIGRATIONS_FAILED",
```

Comment:
I think we should differentiate between being unable to find the migrations and there being a failure with applying them. At the moment we have the same error code for both situations of ERR_MAUTIC_2_MIGRATIONS_FAILED.

```suggestion
                    "ERR_MAUTIC_2_MIGRATIONS_IDENTIFICATION_FAILED",
```

Reply:
I think they should also apply to the M3 migrations

Reply:
Done in https://github.com/mautic/mautic/pull/8824/commits/2751b2787e2bad6c463ef59c056aae2ff281b251